05-agent-demo/risingwave-agent.mdx

```diff
@@ -0,0 +1,626 @@
+---
+
+## Overview
+
+
+
+In this tutorial, you will learn how to implement a custom Anthropic Agent, integrate it with RisingWave MCP, to do simple stream processing with RisingWave.
+
+If you want a reference to the full project you can clone [this](https://github.com/risingwavelabs/awesome-stream-processing/) repository.
+
+## Prerequisites
+
+* Install and run RisingWave. For detailed instructions on how to quickly get started, see the [Quick start](/get-started/quickstart/) guide.
+* Ensure you have an [Anthropic](https://console.anthropic.com/settings/keys) API key.
+* Ensure that the [PostgreSQL](https://www.postgresql.org/docs/current/app-psql.html) interactive terminal, `psql`, is installed in your environment. For detailed instructions, see [Download PostgreSQL](https://www.postgresql.org/download/).
+* Ensure that a Python environment is set up and install the [psycopg2](https://pypi.org/project/psycopg2/) library. 
+* Ensure that you have cloned the [RisingWave MCP](https://github.com/risingwavelabs/risingwave-mcp.git).
+
+## Step 1: Create the file 'agent.py'
+
+```python
+import asyncio
+import os
+from contextlib import AsyncExitStack
+from dataclasses import dataclass
+from anthropic import Anthropic
+from fastmcp import FastMCP
+from abc import ABC, abstractmethod
+from typing import Any, Callable, Tuple
+from mcp import ClientSession, StdioServerParameters
+from mcp.client.sse import sse_client
+from mcp.client.stdio import stdio_client
+```
+
+The MessageHistory class manages the conversation between the user and the agent. It keeps a list of messages, tracks the number of tokens used to stay within the model’s context window limit, and allows messages to be truncated when necessary. It also tracks token usage for each message exchange, formats messages in a structure suitable for Anthropic’s API, and optionally adds cache control headers to recent messages to optimize performance.
+
+```python
+class MessageHistory:
+    """Manages chat history with token tracking and context management."""
+
+    def __init__(
+        self,
+        model: str,
+        system: str,
+        context_window_tokens: int,
+        client: Any,
+        enable_caching: bool = True,
+    ):
+        self.model = model
+        self.system = system
+        self.context_window_tokens = context_window_tokens
+        self.messages: list[dict[str, Any]] = []
+        self.total_tokens = 0
+        self.enable_caching = enable_caching
+        self.message_tokens: list[tuple[int, int]] = (
+            []
+        )  # List of (input_tokens, output_tokens) tuples
+        self.client = client
+
+        # set initial total tokens to system prompt
+        try:
+            system_token = (
+                self.client.messages.count_tokens(
+                    model=self.model,
+                    system=self.system,
+                    messages=[{"role": "user", "content": "test"}],
+                ).input_tokens
+                - 1
```

Comment:
What's this `- 1` for. Seems like a[ magic number](https://en.wikipedia.org/wiki/Magic_number_(programming)).